src/TropicalGeometry/hypersurface.jl

```diff
@@ -248,9 +248,9 @@ end
 
 Return the tropical polynomial used to construct `TropH`.  Raises an error if it is not cached.
 """
-function tropical_polynomial(TropH::TropicalHypersurface)
+function tropical_polynomial(TropH::TropicalHypersurface{minOrMax}) where minOrMax
     @req has_attribute(TropH,:tropical_polynomial) "no tropical polynomial cached"
-    return get_attribute(TropH,:tropical_polynomial)
+    return get_attribute(TropH,:tropical_polynomial)::mpoly_type(TropicalSemiringElem{minOrMax})
```

Comment:
Is there a reason why we use `mpoly_type` instead of `MPolyRingElem` as in the change in line 192?

Reply:
Is in line 192 the coefficient type of this polynomial determined at compile time? if yes, `mpoly_type` should be used there as well

Reply:
Yes: the code in line 192 is too unstable, so the best we can do there is to say "it will be some kind of multivariate polynomial".

There just is no way to specify a more concrete type *by only using the argument types* -- remember, we want the compiler to be able to deduce the types at *compile time*, not at runtime.

At first I thought the result always was a `QQMPolyRingElem` but in one of your book examples it is a polynomial over a rational function field.

As I suggested in the description of this PR already: to improve this, you'd have to add some type parameter(s) to `TropicalHypersurface`, `TropicalLinearSpace`, and possibly others.